### ISSUE TITLE
t macro will not change on locale switch

### ISSUE THREAD

**Issue by moodysalem** [lingui/js-lingui#1078]
**Describe the bug**
when you use the t macro inside a memoized component, the translation text will not change when the locale is switched, unless the component also rerenders

this behavior contrasts the Trans macro which always rerenders on locale switch

**To Reproduce**

create a component that allows you to switch the active locale

set the force render on switch to false

render a pure/memoized component with the t macro

note it does not change when the locale is switched

**Expected behavior**
t is not exported from macro, but rather is accessible via const t = useMacro()

this t should force the component to rerender whenever the locale changes

**Additional context**

adding useLingui() to the top of the memoized component also causes the component to rerender on locale change, which works around the issue, but is not clear

**Comment by vonovak** [comment]
hi, I believe the way to do this currently is

```
const { i18n } = useLingui()

const translation = i18n._(t`string`)
```

but I'm not sure how that is different from 

```
useLingui()

const translation = t`string`
```

perhaps someone more knowledgeable can clarify

**Comment by moodysalem** [comment]
> hi, I believe the way to do this currently is
> 
> 
> 
> ```
> 
> const { i18n } = useLingui()
> 
> 
> 
> const translation = i18n._(t`string`)
> 
> ```
> 
> 
> 
> but I'm not sure how that is different from 
> 
> 
> 
> ```
> 
> useLingui()
> 
> 
> 
> const translation = t`string`
> 
> ```
> 
> 
> 
> perhaps someone more knowledgeable can clarify

i'm pretty sure that would double translate, but something like this would be fine for us as a workaround

```typescript
import { t } from '@lingui/macro'
function useT(): typeof t {
  useLingui()
  return t
}
```

**Comment by semoal** [comment]
Give me tomorrow (already late on Spain) for explaining this correctly, the principal differences between the examples you provided, and in my opinion the correct approach you should take in your project accept's it.

**Comment by jinyang1994** [comment]
Has the problem been fixed? @semoal

**Comment by justindomingue** [comment]
Bump @semoal

**Comment by vonovak** [comment]
@semoal I'd still be interested in your comment on this, thank you :)

**Comment by leikoilja** [comment]
Has it been fixed or explained, @semoal ?

**Comment by micahbule** [comment]
This is still happening on the latest version.

TL;DR: Use `<Trans />` in your React application as much as possible. Explanation below.

Based on observation, it may have something to do with how React works with data changes and how lingui functions in general.

The basic of React component re-rendering relies on two things:
- If any of the prop changes its value
- If its state changes its value

The main difference between the two macros `Trans` and `t` is that `Trans` is utilizing the React Context API to subscribe to changes in the i18n object being passed down in the `<I18nProvider />` during setup of Lingui -- whereas `t` is directly accessing the i18n object from `@lingui/core`. Any changes to the i18n object from `@lingui/core` **will not trigger any re-render** for any components using it.

If you are using the `t` macro in pure components or function components that only display the translated value from props (ex. labels) and you try to switch the activated language, you won't see the changes unless the component re-renders.

**Comment by NinjaTrader-MichaelM** [comment]
This seems like a legitimate issue, why marked as won't fix?

**Comment by crazyyi** [comment]
> This is still happening on the latest version.
> 
> TL;DR: Use `<Trans />` in your React application as much as possible. Explanation below.
> 
> Based on observation, it may have something to do with how React works with data changes and how lingui functions in general.
> 
> The basic of React component re-rendering relies on two things:
> 
> * If any of the prop changes its value
> * If its state changes its value
> 
> The main difference between the two macros `Trans` and `t` is that `Trans` is utilizing the React Context API to subscribe to changes in the i18n object being passed down in the `<I18nProvider />` during setup of Lingui -- whereas `t` is directly accessing the i18n object from `@lingui/core`. Any changes to the i18n object from `@lingui/core` **will not trigger any re-render** for any components using it.
> 
> If you are using the `t` macro in pure components or function components that only display the translated value from props (ex. labels) and you try to switch the activated language, you won't see the changes unless the component re-renders.

I can't get t macro working either on v3.17.2. I have tried `i18n._() `and it seems to generate the same result as `t()` and it actually works on Nextjs production. Why do we need to use `t()` anyway? I am still very confused.
